bolt-modules/prompt/lib/puppet/functions/prompt/menu.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+require 'bolt/error'
+
+# Display a menu prompt and wait for a response. Continues to prompt
+# until an option from the menu is selected.
+#
+# > **Note:** Not available in apply block
+Puppet::Functions.create_function(:'prompt::menu') do
+  # Select from a list of options.
+  # @param prompt The prompt to display.
+  # @param menu A list of options to choose from.
+  # @param options A hash of additional options.
+  # @option options [String] default The default option to return if the user does not provide
+  #   input or if standard in (stdin) is not a tty. Must be an option present in the menu.
+  # @return The selected option.
+  # @example Prompt the user to select from a list of options
+  #   $selection = prompt::menu('Select a fruit', ['apple', 'banana', 'carrot'])
+  # @example Prompt the user to select from a list of options with a default value
+  #   $selection = prompt::menu('Select environment', ['development', 'production'], 'default' => 'development')
+  dispatch :prompt_menu_array do
+    param 'String', :prompt
+    param 'Array[Any]', :menu
+    optional_param 'Hash[String[1], Any]', :options
+    return_type 'Data'
+  end
+
+  # Select from a list of options with custom inputs.
+  # @param prompt The prompt to display.
+  # @param menu A hash of options to choose from, where keys are the input used to select a value.
+  # @param options A hash of additional options.
+  # @option options [String] default The default option to return if the user does not provide
+  #   input or if standard in (stdin) is not a tty. Must be an option present in the menu.
+  # @return The selected option.
+  # @example Prompt the user to select from a list of options with custom inputs
+  #   $menu = { 'y' => 'yes', 'n' => 'no' }
+  #   $selection = prompt::menu('Install Puppet?', $menu)
+  dispatch :prompt_menu do
+    param 'String', :prompt
+    param 'Hash[String[1], Any]', :menu
+    optional_param 'Hash[String[1], Any]', :options
+    return_type 'Data'
```

Comment:
Ah, ok, if we're going to limit it to `Data` (which I think is reasonable) then we should do that in the `menu` parameter and `default` too.

Notably, this excludes sensitive values, but those won't be printed anyway and it's hard to think of a case where you'd want to choose between a bunch of values where they're all `Sensitive [redacted]`.

Reply:
Updated the return type to `Any`. I think there's a compelling use case for selecting targets, and `Target` objects aren't `Data`.

Reply:
Hm, what about `Variant[Target, Data]`? I just don't want to have to think about all the other datatypes and how they'll print to the console. Even Targets I'd say can be strings first and then call `get_target` after choosing.

Reply:
I think requiring targets to be strings and then calling `get_target` after is unnecessarily complicating things for users, especially if they want to pass the name of a group and select from the targets in that group.

As for using `Variant[Target, Data]` over `Any`, I'm not strictly opposed to it... I am concerned that it adds an unnecessary restriction for the function just in the off chance a user wants to select from types that don't print nicely to the console. I think maybe we can discuss with the rest of the team in standup/planning?

Reply:
I think we decided on `Variant[Target, Data]` here for now.